### ISSUE TITLE
Add a way to get source device type from `InputEventAction`

### ISSUE THREAD

**Issue by cgbeutler** [godotengine/godot-proposals#3836]
### Describe the project you are working on

Platformer game.
I have a few spots where I want to filter the input actions based on the device.

### Describe the problem or limitation you are having in your project

`InputEventAction` has a `Device` property for getting the device id that spawned it, but that device id is not unique. There is no way to tell if DeviceId 0 is a joypad, keyboard, mouse, etc.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Add a property for getting the type of device that spawned the action event. Then with that and the id, you have a unique key!

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Maybe add an enum for the different device types that the device id has overlap on. Something like:
```
enum DeviceType { Unknown = 0, Keyboard = 1, Mouse = 2, Gesture = 4, Screen = 8, Joypad = 16, Midi = 32 }
```

Then add a new property to `InputEventAction` that can return that enum type.
```
if actEvent.DeviceType == DeviceType.Joypad and actEvent.Device == curr:
  ...
```

### If this enhancement will not be used often, can it be worked around with a few lines of script?

It could default to unknown if the user creates an event manually but does not set the device type. (Thus ignoring it would be fine, if they don't need it.)

### Is there a reason why this should be core and not an add-on in the asset library?

Unless I have missed something, there is currently no way to truly map an `InputEventAction` back to the source device. That seems like a core oversight to me...?

**Comment by KoBeWi** [comment]
Why do you need to know the device though?

**Comment by cgbeutler** [comment]
If you are looking to provide workarounds or alternatives, I have already implemented those and don't need help there.

I mostly request this feature because it would really simplify a lot of action-based work.
Other `InputEvent`s have a way to figure out the exact device because they are unique types based on those devices. This info is lost as an event is sliced down to an action event. It seems like an oversight that InputEventAction is the only one that can't determine an exact device of origin.
I could see maybe allowing InputEventAction to just be given an 'internal event' type field that can hold the upstream event that was turned into the action event. Then I could at least get that event and do a lookup that way.

In case you weren't asking as a way to help, but actually couldn't see why this would be nice to have, here are a few examples of where it would help me:
I would like to define a reduced/minimal set of actions marked with "any joypad" so that an unbounded number of players could join. Inputs could then be filtered on controller type/number. Currently, you would have to use a more roundabout workaround using action lookups on other events or looking up actions in a non-event way.
I also have some "big brother/little brother" systems in place. This proposal would make is _super_ easily filter inputs on the start menu, allowing the big bro device to control it, but nothing else.
The reduced set action binding would also allow something like watching for the "ui_accept" event from any device, then lookup who hit it, marking that player as 'ready'.

**Comment by KoBeWi** [comment]
> If you are looking to provide workarounds or alternatives, I have already implemented those and don't need help there.

No, you just said that the feature is lacking without providing any use-case.

The most common way to support inputs for many players is prefixing your actions with e.g. `p1_`, `p2_` etc. and for each set of actions use a different device id. The worst part is setup, after that it's not different than manually checking device - just use a different action name based on player.

#3555 is probably related.

**Comment by simast** [comment]
I just hit a similar issue trying to implement a simple "Zoom In" action. Imagine you want to have a single action with the following bindings:

```
Zoom In
  - Mouse Wheel Up
  - Keyboard Page Up
  - Joypad Right Trigger
```

And when this `InputEventAction` triggers I would prefer to implement two completely different approaches for zooming:
1. For mouse/keyboard input apply some arbitrary zoom amount (with tween animation etc).
2. For analog joypad input apply the strength directly as zoom.

As I understand right now there is no good way to implement this as a single input action.

**Comment by laurentsenta** [comment]
Found this issue looking for the same feature,

My use case:
- I want to provide keyboard, mouse, and pad controls (eventually touch screen too)
- I want to show UI feedback based on the last used type of input
  - Example: "Press (Y) to interact" or "Press [Enter] to Interact"

Since `input.device` is `0` for all my inputs, I believe my only option here is to write a huge `if type of input is ... then ... elif type is ... `.

That's a workaround, but it breaks a bunch of expectations:
- Since there is already an abstraction above the event (`event.is_action_pressed`, etc.), I was expecting to find something similar on the input type (`event.is_keyboard()`, etc.).
- At least I expected the `input.device` to be different for a mouse, keyboard, pad, etc.

**Comment by soloparatolos** [comment]
I'm just using one set of actions in my game and getting the ID to know who pushed the button. I mainly target controllers but while looking at adding keyboard support I just realized that the ID is the same for keyboard and joypad one. This looks conceptually weird to me. 

In my head, if an inputEvent happens, the Device ID should idetify the device unequivocally... but thats not the case. 

I guess the only solution reamain implementing exta steps; duplicating actions or extra checks for the input type.

The proposed solution would work for me. Dont know if remaping device ids would be doable, but for me it would do the trick too.

**Comment by Nintorch** [comment]
Hello, it seems that `InputEventAction` is not emitted by the engine ( https://docs.godotengine.org/en/stable/classes/class_inputeventaction.html#description ), so I think it's the user who's responsible to set the `device` property of the event.
See also https://github.com/godotengine/godot/pull/116274
